### ISSUE TITLE
How can I support join conditions for deeply nested has-many and many-to-many relationships?

### ISSUE THREAD

**Issue by matthew-noken** [go-pg/pg#952]
Hi,

I have some really complex queries that in the most extreme example require up to 50 joins. I want to support soft deletes via a `deleted_at` field (for all tables), but I don't believe there's any reasonable way to do that on the query side with pg.

If these were all has-one relationships, it would be easy enough, though tedious and unwieldy. I could call `Where("table_name"."deleted_at" IS NULL)` for each joined table.

The problem is when there are has-many or many-to-many relationships, and these are nested, and multiple queries are generated. In simple cases, I can't just call `Where`, as some queries will have those tables in the join and some won't. Calling `Relation` also "works" in some cases, but the query that's generated only seems to apply to the primary table against which it's called, so a child that joins against grandchildren will only have the conditions applied to the child, not the grandchildren.

Ultimately this becomes a game of "guess the query plan," but even doing that it seems like an intractable problem as the library currently exists.

Here are three solutions to this problem that I can think of, from worst to best (from my perspective):

1) Allow me to specify a readTableName and a writeTableName on the model. Then I can create views for every table that have my where condition applied, and pg can simply choose the correct table based on whether it's a select or an insert, update, or delete. This is tedious to set up, but only needs to happen once.

2) Allow me to specify a `pg:selectJoinWhere` (or similar) property on the model, where the value is a condition for all select statements against the table (and perhaps optionally updates and deletes as well). It would need to allow me to specify a placeholder for the table name, since that can change depending on the query.

3) Allow me to designate a column as a soft-delete column, such as `pg:delete`. The library takes care of the rest unless a method is called telling it to do otherwise (e.g., `WithDeleted`).

I'd be willing to pay for this feature if it will help accelerate development.

**Comment by vmihailenco** [comment]
Hi,

Do you know about 2nd arg in `Relation`, right?

```
Relation("Join", func(q *orm.Query) (*orm.Query, error) {
    q = q.Where("?TableAlias.deleted_at IS NULL")
    return q, nil
})
```

It is still tedious to do if you have lots of joins but possible.

With https://github.com/go-pg/pg/pull/953 merged you should be able to do following

// BeforeSelectQuery is called before Select query is executed.
func (*Book) BeforeSelectQuery(db orm.DB, q *orm.Query) (*orm.Query, error) {
    q = q.Where("?TableAlias.deleted_at IS NULL")
    return q, nil
}

Which I think should be enough for your use case.

>Allow me to specify a readTableName and a writeTableName on the model.

You can create `ReadModel` and `WriteModel` and use them accordingly. I guess this is too much work too, but occasionally (once so far :) I use such approach myself. With embedding this should not be too bad:

```go
type Model struct{}

type ReadModel struct {
    tableName
    Model
}
```

>Allow me to designate a column as a soft-delete column, such as pg:delete

Let me think about this for few days. It does not look like it is too hard to add, but I don't have much experience with soft deletes.

>I'd be willing to pay for this feature if it will help accelerate development.

Thanks - may be next time :)

**Comment by matthew-noken** [comment]
Sorry for so many replies, but I've found another issue when attempting this solution. When your model is named `ReadResident`, the query for `Secrets` in the above example won't work. It appears to be a problem with has-many relationships in that they are based on name.

Let's say you set your column call as `Column("read_resident.*", "Secrets")`; the query will generate `SELECT "Secrets" FROM ...`, treating "Secrets" as a column, which creates an error.

With deeply nested joins, it will even flatten the list against a property. For example, if I have `Column("song.*", "Do", "Do.Re", "Do.Re.Me", "Do.Re.Me.Fa", "Do.Re.Me.Fa.So")`, and Re is a has-many relationship, it will (incorrectly) generate `SELECT "do"."Re", "do.Me", "do.Fa", "do.So" FROM ... AS "song" LEFT JOIN ... AS "do" ...`.

Now, if I use the original model (`Resident`), everything works fine. Identical in every way except the name (formerly `ReadResident`), the type of `Secrets` (formerly `ReadSecrets`), and the table name designated via `tableName`.

I searched through the wiki and the code for a tag to force it to identify the correct column, but I couldn't find one that worked.

What I've done to resolve this situation is to make all my read models—the ones I join against—named the default, and all my write models named something else (e.g., `WriteResident`, `WriteSecret`). I cannot join any has-many relationship with the write models, which is fine for my purposes.

Note that when updating the model with the new setup (`WriteResident`, etc.) I get the same error as above, but in reverse. To illustrate this,

```
var resident WriteResident
err := db.Model(&resident).Where("id = ?", id).First()
if err != nil {
    return err
}
resident.Name = "Barbara"
return db.Update(&resident)
```

This will generate a SQL error like "ERROR: column write_resident.secrets does not exist at character..."

However, by changing it to the following, everything works as expected:

```
return db.
    Model(&WriteResident{}).
    Set("name = ?", "Barbara").
    Where("id = ?", id).
    Update()
```

So to recap, I have been able to find a workaround to the issues around has-many by doing the following:

1. Create views for all tables where `deleted_at` is null
2. Point your existing models at the new views
3. Create duplicate models named something like `WriteResident` pointing at the tables
4. Do not query with joins against any `Write`-prefixed models
5. Do not use the `db.Update(&resident)`-style functionality

**Comment by vmihailenco** [comment]
https://github.com/go-pg/pg/pull/958 added the ability to specify different table name for select queries which I guess should be helpful here:

```go
type Model struct {
    tableName struct{} `sql:"select:view_name"`
}
```

I have plans to work more on this issue but later this week.

**Comment by vmihailenco** [comment]
https://github.com/go-pg/pg/pull/961 added support for soft deletes. See https://godoc.org/github.com/go-pg/pg#example-DB-Model-SoftDelete for an example. It seems to work, but I did not spend much time thinking about corner cases and reserve right to break existing behavior.

**Comment by matthew-noken** [comment]
Thanks for your work on this! I just returned from vacation and am reviewing the change. It looks like it works correctly for select and update (it checks `deleted_at` in those cases where a field is optional), but not delete. But very close!

When I run a delete on a joining table (not a populated record), `resident_secrets`, I see the following behavior:

Go:
```
_, err = db.
	Model(&ResidentSecret{}).
	Where("resident_id = ?", id).
	Where("secret_id IN (?)", pg.In(secretIDs)).
	Delete()
```

SQL:
```
SELECT count(*) FROM residents AS "resident" WHERE (id = 3) AND "resident".deleted_at IS NULL

UPDATE resident_secrets AS "resident_secret" SET "created_at" = '0001-01-01 00:00:00+00:00:00', "updated_at" = '0001-01-01 00:00:00+00:00:00', "deleted_at" = '2018-09-04 15:22:18.633358014+00:00:00', "resident_id" = 0, "secret_id" = 0 WHERE (resident_id = 3) AND (secret_id IN (2)) AND "resident_secret".deleted_at IS NULL
```

Of course we don't want to reset all fields in the record (and that violates FK constraints in any event).

**Comment by vmihailenco** [comment]
Thanks for the report - that should be fixed in master.

**Comment by matthew-noken** [comment]
Perfect, I think this solves our issue. I'll create a new ticket if any related issues come up. Thank you for your work on this!
